### PR TITLE
AWS EXPANDR 5778

### PR DESCRIPTION
**Pull request by johnnywilkes**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://jira-dc.paloaltonetworks.com/browse/EXPANDR-5778

## Description
- Added 3 commands necessary to find out what AWS EC2 instance is using a public IP in an organization with minimal API calls:
  - aws-ec2-describe-ipam-resource-discoveries - Describes IPAM resource discoveries. A resource discovery is an IPAM component that enables IPAM to manage and monitor resources that belong to the owning account.
  - aws-ec2-describe-ipam-resource-discovery-associations - Describes resource discovery association with an Amazon VPC IPAM. An associated resource discovery is a resource discovery that has been associated with an IPAM.
  - aws-ec2-get-ipam-discovered-public-addresses - Gets the public IP addresses that have been discovered by IPAM.
  - 
## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 1]
```diff
@@ -3805,7 +3805,87 @@ script:
...
+      name: AddressRegion
+    outputs:
+    - contextPath: AWS.EC2.IpamResourceDiscoveries.IpamResourceDiscoveryId
+      description: The resource discovery ID.
+      type: String
+    - contextPath: AWS.EC2.IpamResourceDiscoveries.OwnerId
+      description: The ID of the owner.
+      type: String
+    - contextPath: AWS.EC2.IpamResourceDiscoveries.IpamResourceDiscoveryRegion
+      description: The resource discovery Region.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3827]:
```suggestion
      description: The resource discovery region.
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 2]
```diff
@@ -3805,7 +3805,87 @@ script:
...
+  - arguments:
+    - description: A comma-separated list of the IPAM resource discovery IDs.
+      name: IpamResourceDiscoveryIds
+    - description: One or more filters Separated by ';'. See AWS documentation for details & filter options.
+      name: Filters
+    - description: The maximum number of results to return in a single call. Specify a value between 5 and 1000.
+      name: MaxResults
+    - description: The token for the next set of results.
+      name: NextToken
+    - description: The Amazon Web Services Region for the IP address.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3817]:
```suggestion
    - description: The Amazon Web Services region for the IP address.
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 3]
```diff
@@ -3805,7 +3805,87 @@ script:
       name: roleSessionDuration
     description: Release on demand dedicated host.
     name: aws-ec2-release-hosts
-  dockerimage: demisto/boto3py3:1.0.0.76862
+  - arguments:
+    - description: A comma-separated list of the IPAM resource discovery IDs.
+      name: IpamResourceDiscoveryIds
+    - description: One or more filters Separated by ';'. See AWS documentation for details & filter options.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3811]: Please add a link to the documentation.

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 4]
```diff
@@ -3805,7 +3805,87 @@ script:
...
+      type: String
+    - contextPath: AWS.EC2.IpamResourceDiscoveries.IpamResourceDiscoveryRegion
+      description: The resource discovery Region.
+      type: String
+    description: Describes IPAM resource discoveries. A resource discovery is an IPAM component that enables IPAM to manage and monitor resources that belong to the owning account.
+    name: aws-ec2-describe-ipam-resource-discoveries
+  - arguments:
+    - description: A comma-separated list of the resource discovery association IDs.
+      name: IpamResourceDiscoveryAssociationIds
+    - description: One or more filters Separated by ';'. See AWS documentation for details & filter options.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3834]: Please add a link to the documentation.

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 5]
```diff
@@ -3805,7 +3805,87 @@ script:
...
+      name: Filters
+    - description: The maximum number of results to return in a single call. Specify a value between 5 and 1000.
+      name: MaxResults
+    - description: The token for the next set of results.
+      name: NextToken
+    - description: The Amazon Web Services Region for the IP address.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3840]:
```suggestion
    - description: The Amazon Web Services region for the IP address.
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 6]
```diff
@@ -3805,7 +3805,87 @@ script:
...
+      name: AddressRegion
+    outputs:
+    - contextPath: AWS.EC2.IpamResourceDiscoveryAssociations.IpamResourceDiscoveryAssociationId
+      description: The resource discovery association ID.
+      type: String
+    - contextPath: AWS.EC2.IpamResourceDiscoveryAssociations.IpamResourceDiscoveryId
+      description: The resource discovery ID.
+      type: String
+    - contextPath: AWS.EC2.IpamResourceDiscoveryAssociations.IpamRegion
+      description: The IPAM home Region.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3850]:
```suggestion
      description: The IPAM home region.
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 7]
```diff
@@ -3805,7 +3805,87 @@ script:
...
+      type: String
+    description: Describes resource discovery association with an Amazon VPC IPAM. An associated resource discovery is a resource discovery that has been associated with an IPAM.
+    name: aws-ec2-describe-ipam-resource-discovery-associations
+  - arguments:
+    - description: An IPAM resource discovery ID.
+      name: IpamResourceDiscoveryId
+      required: true
+    - description: The Amazon Web Services Region for the IP address.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3858]:
```suggestion
    - description: The Amazon Web Services region for the IP address.
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml [hunk 8]
```diff
@@ -3805,7 +3805,87 @@ script:
...
+      name: AddressRegion
+      required: true
+    - description: One or more filters Separated by ';'. See AWS documentation for details & filter options.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/AWS-EC2.yml:3861]: Please add a link to the documentation.

File: Packs/AWS-EC2/Integrations/AWS-EC2/README.md [hunk 9]
```diff
@@ -3237,3 +3237,247 @@ There is no context output for this command.
...
+#### Base Command
+
+`aws-ec2-describe-ipam-resource-discoveries`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| IpamResourceDiscoveryIds | A comma-separated list of the IPAM resource discovery IDs. | Optional | 
+| Filters | One or more filters Separated by ';'. See AWS documentation for details &amp; filter options. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/README.md:3255]: Please add a link to the documentation.

File: Packs/AWS-EC2/Integrations/AWS-EC2/README.md [hunk 10]
```diff
@@ -3237,3 +3237,247 @@ There is no context output for this command.
...
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| IpamResourceDiscoveryIds | A comma-separated list of the IPAM resource discovery IDs. | Optional | 
+| Filters | One or more filters Separated by ';'. See AWS documentation for details &amp; filter options. | Optional | 
+| MaxResults | The maximum number of results to return in a single call. Specify a value between 5 and 1000. | Optional | 
+| NextToken | The token for the next set of results. | Optional | 
+| AddressRegion | The Amazon Web Services Region for the IP address. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/README.md:3258]:
```suggestion
| AddressRegion | The Amazon Web Services region for the IP address. | Optional | 
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/README.md [hunk 11]
```diff
@@ -3237,3 +3237,247 @@ There is no context output for this command.
...
+| NextToken | The token for the next set of results. | Optional | 
+| AddressRegion | The Amazon Web Services Region for the IP address. | Optional | 
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| AWS.EC2.IpamResourceDiscoveries.IpamResourceDiscoveryId | String | The resource discovery ID. | 
+| AWS.EC2.IpamResourceDiscoveries.OwnerId | String | The ID of the owner. | 
+| AWS.EC2.IpamResourceDiscoveries.IpamResourceDiscoveryRegion | String | The resource discovery Region. |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/README.md:3266]:
```suggestion
| AWS.EC2.IpamResourceDiscoveries.IpamResourceDiscoveryRegion | String | The resource discovery region. | 
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/README.md [hunk 12]
```diff
@@ -3237,3 +3237,247 @@ There is no context output for this command.
...
+#### Base Command
+
+`aws-ec2-describe-ipam-resource-discovery-associations`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| IpamResourceDiscoveryAssociationIds | A comma-separated list of the resource discovery association IDs. | Optional | 
+| Filters | One or more filters Separated by ';'. See AWS documentation for details &amp; filter options. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/README.md:3364]: Please add a link to the documentation.

File: Packs/AWS-EC2/Integrations/AWS-EC2/README.md [hunk 13]
```diff
@@ -3237,3 +3237,247 @@ There is no context output for this command.
...
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| IpamResourceDiscoveryAssociationIds | A comma-separated list of the resource discovery association IDs. | Optional | 
+| Filters | One or more filters Separated by ';'. See AWS documentation for details &amp; filter options. | Optional | 
+| MaxResults | The maximum number of results to return in a single call. Specify a value between 5 and 1000. | Optional | 
+| NextToken | The token for the next set of results. | Optional | 
+| AddressRegion | The Amazon Web Services Region for the IP address. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/README.md:3367]:
```suggestion
| AddressRegion | The Amazon Web Services region for the IP address. | Optional | 
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/README.md [hunk 14]
```diff
@@ -3237,3 +3237,247 @@ There is no context output for this command.
...
+| NextToken | The token for the next set of results. | Optional | 
+| AddressRegion | The Amazon Web Services Region for the IP address. | Optional | 
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| AWS.EC2.IpamResourceDiscoveryAssociations.IpamResourceDiscoveryAssociationId | String | The resource discovery association ID. | 
+| AWS.EC2.IpamResourceDiscoveryAssociations.IpamResourceDiscoveryId | String | The resource discovery ID. | 
+| AWS.EC2.IpamResourceDiscoveryAssociations.IpamRegion | String | The IPAM home Region. |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/README.md:3375]:
```suggestion
| AWS.EC2.IpamResourceDiscoveryAssociations.IpamRegion | String | The IPAM home region. | 
```

File: Packs/AWS-EC2/Integrations/AWS-EC2/README.md [hunk 15]
```diff
@@ -3237,3 +3237,247 @@ There is no context output for this command.
...
+
+`aws-ec2-get-ipam-discovered-public-addresses`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| IpamResourceDiscoveryId | An IPAM resource discovery ID. | Required | 
+| AddressRegion | The Amazon Web Services Region for the IP address. | Required | 
+| Filters | One or more filters Separated by ';'. See AWS documentation for details &amp; filter options. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/Integrations/AWS-EC2/README.md:3425]: Please add a link to the documentation.

File: Packs/AWS-EC2/ReleaseNotes/1_2_41.md [hunk 16]
```diff
@@ -0,0 +1,9 @@
+
+#### Integrations
+
+##### AWS - EC2
+
+- Added 3 commands:
+  - ***aws-ec2-describe-ipam-resource-discoveries***
+  - ***aws-ec2-describe-ipam-resource-discovery-associations***
+  - ***aws-ec2-get-ipam-discovered-public-addresses***
```
Comment (ShirleyDenkberg) [inline Packs/AWS-EC2/ReleaseNotes/1_2_41.md:9]:
```suggestion
Added 3 commands:
- ***aws-ec2-describe-ipam-resource-discoveries***
- ***aws-ec2-describe-ipam-resource-discovery-associations***
- ***aws-ec2-get-ipam-discovered-public-addresses***
```

**Comment by ShirleyDenkberg** [comment]
@amshamah419 Doc review completed.

**Comment by johnnywilkes** [comment]
@amshamah419 / @RotemAmit , any ideas how to get rid of this pre-commit/lint error?  I will share demo video with you via google drive.  Keep in mind that I will be away until Monday December 11th.  Please contact @BigEasyJ or @capanw for assistance

@BigEasyJ , can you keep track of this PR while I am away, please?

**Comment by RotemAmit** [comment]
> @amshamah419 / @RotemAmit , any ideas how to get rid of this pre-commit/lint error? I will share demo video with you via google drive. Keep in mind that I will be away until Monday December 11th. Please contact @BigEasyJ or @capanw for assistance
> 
> @BigEasyJ , can you keep track of this PR while I am away, please?

The aws_session is the one creating the EC2Client, so you should move this part
client = aws_client.aws_session(
        service='ec2',
        region=args.get('AddressRegion') or args.get('region'),
        role_arn=args.get('roleArn'),
        role_session_name=args.get('roleSessionName'),
        role_session_duration=args.get('roleSessionDuration'),
    )
to the main function and then send it to the relevant command for it to "know" the EC2Client. It means that you can do it only once in the main instead of in every command function.

**Comment by amshamah419** [comment]
@johnnywilkes @BigEasyJ The PR looks great, can you all take a look at the unit tests? Afterwards, I think we can merge.

**Comment by BigEasyJ** [comment]
@amshamah419 I am looking into this for a resolution. For some reason @RotemAmit's suggest is breaking some tests where the `client` still is pointing to `EC2Client` class however it can not access any of it's functions like previously before.

**Review by amshamah419 — APPROVED** [review]
Great job!

**Comment by BigEasyJ** [comment]
@amshamah419 this PR is ready

**Comment by johnnywilkes** [comment]
@amshamah419 , this is ready, please merge when possible

**Pull request merged**